### ISSUE TITLE
Should we escape slashs on path?

### ISSUE THREAD

**Issue by Kludex** [tiangolo/fastapi#1681]
### First check

* [X] I added a very descriptive title to this issue.
* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [X] I already read and followed all the tutorial in the docs and didn't find an answer.
* [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [X] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description

This is more a question/feature request. Currently, we are not able to withstand forward slashes in the `Path`. Do we want to have that available?

### Minimal example

Having those two endpoints:

```python
from fastapi import FastAPI

app = FastAPI()

@app.post("/{path}")
def post_method(path: str):
    return "ok"

@app.post("/{path}/{another}")
def post_method_another(path: str, another: str):
    return "not ok"
```

I want to be able to access the first endpoint with the following request:
`curl -X POST "http://localhost:8000/hahahaha%2Fahhaa" -H "accept: application/json" -d ""`

The `%2F` is known way to escape forward slashes. 

### Additional context

This issue was brought by @hang15 on our gitter and on [StackOverflow](https://stackoverflow.com/questions/62770461/starlette-fastapi-route-path-components-with-forward-slash).

**Comment by tiangolo** [comment]
Yeah, I think this [StackOverflow](https://stackoverflow.com/a/42487180/219530) answer has good info pointing to the W3C recommendations which would be sort of the authority (I don't think the URL RFC have something specific about that). 

So I think that *should* be the behavior. Nevertheless, this is actually handled in Starlette directly, not in FastAPI. But it would be a valid Starlette bug/feature request.

**Comment by ycd** [comment]
Well, we can actually achieve that by creating a custom route, 

```python
from fastapi import FastAPI, Body, APIRouter, Request, Response
from typing import Callable
from fastapi.routing import APIRoute
from pydantic import BaseModel
import json
from pprint import pprint as print


class CustomRoute(APIRoute):
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            old_scope = request.scope["raw_path"].decode("utf-8")
            new_scope = old_scope.replace("%2F", "/").encode("utf-8")
            request.scope["raw_path"] = new_scope


            response = await original_route_handler(request)

            return response

        return custom_route_handler


app = FastAPI()
router = APIRouter(route_class=CustomRoute)


class TestModel(BaseModel):
    name: str


@router.post("/{some}/{other_path}")
async def dummy(some: str, other_path: str):
    return {
        "some": some,
        "other_path": other_path,
    }


app.include_router(router)
```

This changes the `raw_path` on fly

```python
curl -X POST "http://127.0.0.1:8000/asdf%2Fasdf" -H  "accept: application/json" -d ""
{"some":"asdf","other_path":"asdf"}
```

**Comment by SamDM** [comment]
The answer posted by `ycd` does not seem to work, and does not seem to answer the original question either.
I.e., given the endpoint:

```python
@app.post("/{path}")
def post_method(path: str):
    return "ok"
```

You are not able to call it with: 
```
curl -X POST "http://localhost:8000/hahahaha%2Fahhaa" -H "accept: application/json" -d ""
```

I've put logging statements in the `custom_route_handler` above and this code is not being called when the path component contains an escaped slash `%2F`. A 404 is send back before even reaching that point in the code.
